### ISSUE TITLE
some bugs about the new `warning` feature

### ISSUE THREAD

**Issue by virusdefender** [teal-language/tl#285]
@euclidianAce 

## case 1

```
local os = require("os")
print(os.clock())
```

> redeclaration of variable 'os'

there is no another `os` variable?

## case 2

```
local f = function(event: number)
    if event == 1 then
        goto break_restart
    else
        return
    end
    
    ::break_restart::
    print("break")
end

print(f(1))
print(f(2))
```

> test.tl:3:9: unused variable @unresolved: table: 0x55b6e9883a90
> est.tl:3:9: unused variable @unresolved: table: 0x55b6e9883a90
> test.tl:8:5: unused label ::break_restart::
> test.tl:3:9: unused variable @unresolved: table: 0x55b6e9883a90

**Comment by euclidianAce** [comment]
thanks for the report, will look into these when I get the chance!
> there is no another os variable?

Technically there is the global `os` variable that you shadow with the local one, but this is a decently common pattern that this should probably be an exception. A small workaround for now if you want to get rid of the warning would just be to make a local of the function you need like so:
```
local os_clock = os.clock
print(os_clock)
```

As for the second case, I'm not exactly sure how labels are resolved, but its hopefully an easy solution. (pinging @hishamhm since you added the warnings for labels)

**Comment by euclidianAce** [comment]
okay, #286 should solve that second case. The first one is a bit more complex of an issue that might warrant some discussion

**Comment by virusdefender** [comment]
another case 

```
local start,  _, _ = string.find(s, sub, 1, true)
```

```
redeclaration of variable '_' (originally declared at xx:xx)
```

should it be accepted?

it's allowed in golang

```
package main

import "fmt"

func a() (int, int, string) {
	return 1, 2, "test"
}

func main() {
	val, _, _ := a()
	fmt.Println(val)
}
```

**Comment by euclidianAce** [comment]
I don't know much about golang, but after a cursory glance the compiler throws an error about `a` returning 3 values so you need placeholder values there to acknowledge the arity of the function. In Lua you can just ignore those values without a placeholder
```lua
local function a()
   return 1, 2, "test"
end

local val = a()
print(val)
```
In fact it appears that golang is one of many languages that actually treats `_` as sort of a black hole that is used to acknowledge the existence of values, but not use them. Whereas in languages like Lua (and consequently Teal) just use `_` as convention for something to be ignored, e.g. in Go:
```go
package main
import "fmt"
func main() {
	_ := 10
	fmt.Println(_)
}
```
```
$ go run main.go
./main.go:4:4: no new variables on left side of :=
./main.go:5:13: cannot use _ as value
```
whereas in Lua this is perfectly fine:
```lua
local _ = 10
print(_)
```
```
$ lua main.lua
10
```

TL;DR: The warning (imho) is justified as ignoring extra return values just by not assigning them to anything seems more idiomatic to Lua.

**Comment by catwell** [comment]
Wouldn't you get the same warning if you did this?

```lua
local _, _, x = f()
```

Personally I have a rule to ignore this case in my `.luacheckrc` and I think it should be the default in tl.

Yes, you can work around it (for instance using `select`) but the use of _ is kind of idiomatic.

**Comment by euclidianAce** [comment]
huh, didn't think about that. Yeah, that warning should probably be ignored in that case

**Comment by hishamhm** [comment]
Agree, in the realm of warnings handling idioms is fair game — you're already doing `_unused` for ignoring unused variables, so `_` to mean a "don't care" variable could exempt it from redeclaration warnings. :+1:

**Comment by hishamhm** [comment]
I think with #300 and #286 in, this can be closed?

**Comment by euclidianAce** [comment]
Just pushed a small adjustment to ignore redeclaration of vars prefixed with '_', so I'm confident in this being closed, feel free to open another issue/reopen this if something else pops up
